docs/WritingADeviceDriver.md

```diff
@@ -0,0 +1,220 @@
+Writing a CHERIoT Device Driver
+===============================
+
+CHERIoT aims to be small and easy to customize.
+It does not have a generic device driver interface but it does have a number of tools that make it possible to write modular device drivers.
+
+What is a device?
+-----------------
+
+From the perspective of the CPU, a device is something that you communicate with via a memory-mapped I/O interface, which may (optionally) generate interrupts.
+There are several devices that the core parts of the RTOS interact with:
+
+ - The UART, which is used for writing debug output during development.
+ - The core-local interrupt controller, which is used for managing timer interrupts.
+ - The platform interrupt controller, which is used for managing external interrupts.
+ - The revoker, which scans memory for dangling capabilities (pointers) and invalidates them.
+
+Specifying a device's locations
+-------------------------------
+
+Devices are specified in the [board description file](BoardDescriptions.md).
+The two relevant parts are the `devices` node, which specifies the memory-mapped I/O devices and the `interrupts` section that describes how external interrupts should be configured.
+For example, our initial FPGA prototyping platform had sections like this describing its Ethernet device:
+
+```json
+    "devices" : {
+        "ethernet" : {
+            "start" : 0x98000000,
+            "length": 0x204
+        },
+        ...
+    },
+    "interrupts": [
+        {
+            "name": "Ethernet",
+            "number": 16,
+            "priority": 3
+        }
+    ],
+```
+
+The first part says that the ethernet device's MMIO space is 0x204 bytes, starting at address 0x98000000.
+The second says that interrupt number 16 is used for the ethernet device.
+
+Accessing the memory-mapped I/O region
+--------------------------------------
+
+The `MMIO_CAPABILITY` macros is used to get a pointer to memory-mapped I/O devices.
+This takes two arguments.
+The first is the C/C++ type of the pointer, the second is the name from the board configuration file.
+For example, to get a pointer to the memory-mapped I/O space for the ethernet device above, we might do something like:
+
+```c
+struct EthernetMMIO
+{
+    // Control register layout here:
+    ...
+};
+
+volatile struct EthernetMMIO *ethernet_device()
+{
+    return MMIO_CAPABILITY(struct EthernetMMIO, ethernet);
+}
+```
+
+Note that this macro must be used in code, it cannot be used for static initialisation.
+The macro expands to a load from the compartment's import table and so there is no point assigning the result to a global: you will get smaller code using it directly.
+
+Now that you have a `volatile` pointer to the device's MMIO region, you can access its control registers directly.
```

Comment:
I hate to nit pick but isn't it the struct that is volatile, not the pointer?

Reply:
```suggestion
Now that you have a pointer to a `volatile` object representing the device's MMIO region, you can access its control registers directly.
```